[+Doc] Troubleshooting / Hot Spotting (#95429)
elastic/elasticsearch: docs/reference/troubleshooting/common-issues/hotspotting.asciidoc

spotting:

* Resources are allocated non-uniformly. For example, if one hot node is 
given half the cpu of its peers. {es} expects all nodes on a 
<<data-tiers,data tier>> to share the same hardware profiles or 
specifications.

* Resources are consumed by another service on the host, including other 
{es} nodes. Refer to our <<dedicated-host,dedicated host>> recommendation.

* Resources experience different network or disk throughputs. For example, if one 
node's I/O is lower than its peers. Refer to 
<<tune-for-indexing-speed,Use faster hardware>> for more information.

* A JVM that has been configured with a heap larger than 31GB. Refer to <<set-jvm-heap-size>> 
for more information.

* Problematic resources uniquely report <<setup-configuration-memory,memory swapping>>. 

[discrete]
[[causes-shards]]
===== Shard distributions

{es} indices are divided into one or more link:{wikipedia}/Shard_(database_architecture)[shards] 
which can sometimes be poorly distributed. {es} accounts for this by <<modules-cluster,balancing shard counts>> 
across data nodes. As link:{blog-ref}whats-new-elasticsearch-kibana-cloud-8-6-0[introduced in version 8.6], 

> {es} by default also enables <<modules-cluster,desired balancing>> to account for ingest load; 
> though there's always link:{es-repo}/issues/17213[room for improvement]. 

Reviewer: ```suggestion
{es} by default also enables <<modules-cluster,desired balancing>> to account for ingest load.. 
```

Reviewer: Is the "room for improvement" note helpful?

Author: It links to the parent discussion about why shard hot spotting occurs at all in Elasticsearch vs our attempts to account for it. I think it's helpful for advanced users, but maybe not newer. What do you think?

Reviewer: I'm not sure it belongs in documentation. A Github conversation is ephemeral, and who knows if it will still be relevant a year from now, for example. But if you feel strongly, we can keep it, of course.

A node may still experience hot spotting either due to write-heavy indices or by the 
overall shards it's hosting.

[discrete]
[[causes-shards-nodes]]
====== Node level

You can check for shard balancing via <<cat-allocation,cat allocation>>, though as of version 
8.6, <<modules-cluster,desired balancing>> may no longer fully expect to 
balance shards. Kindly note, both methods may temporarily show problematic imbalance during 
<<cluster-fault-detection,cluster stability issues>>.

For example, let's showcase two separate plausible issues using cat allocation:

[source,console]
----
GET _cat/allocation?v&s=node&h=node,shards,disk.percent,disk.indices,disk.used
----

Which could return:

[source,console-result]
----
node   shards disk.percent disk.indices disk.used
node_1    446           19      154.8gb   173.1gb
node_2     31           52       44.6gb   372.7gb
node_3    445           43      271.5gb   289.4gb
----
// TEST[skip:illustrative response only]

> Here we see two significantly unique situations. The first is a node 
> recently restarted so has much lower shards than all other nodes, 
> which also relates to `disk.indices` being much smaller than `disk.used` 
> while shards are recovering as seen via <<cat-recovery,cat recovery>>. The 
> second is one node with par shards shows much higher `disk.percent`. This 
> occurs when either shards are not evenly sized (refer to 
> <<shard-size-recommendation>>) or when there's a lot of empty indices.

Reviewer: ```suggestion
Here we see two significantly unique situations. `node_1` has 
recently restarted, so it has a much lower number of shards than all other nodes. 
This also relates to `disk.indices` being much smaller than `disk.used` 
while shards are recovering as seen via <<cat-recovery,cat recovery>>. The 
second situation is that `node_3` has a higher `disk.percent` than `node_1`, 
even though they hold roughly the same number of shards. This 
occurs when either shards are not evenly sized (refer to 
<<shard-size-recommendation>>) or when there are a lot of empty indices.
```

Reviewer: I think referring to the nodes by name makes it a bit more clear.

You can confirm hot spotted nodes via <<cluster-nodes-stats,the nodes stats API>>, 
potentially polling twice over time to only checking for the stats differences 
between them rather than polling once giving you stats for the node's 
full <<cluster-nodes-usage,node uptime>>. For example, to check all nodes 
indexing stats:

[source,console]
----
GET _nodes/stats?human&filter_path=nodes.*.name,nodes.*.indices.indexing
----

[discrete]
[[causes-shards-index]]
====== Index level

Hot spotted nodes frequently surface via <<cat-thread-pool,cat thread pool>>'s 
`write` and `search` queue backups. For example:

[source,console]
----
GET _cat/thread_pool/write,search?v=true&s=n,nn&h=n,nn,q,a,r,c
----

Which could return:

[source,console-result]
----
n      nn       q a r    c
search node_1   3 1 0 1287
search node_2   0 2 0 1159
search node_3   0 1 0 1302
write  node_1 100 3 0 4259
write  node_2   0 4 0  980
write  node_3   1 5 0 8714
----
// TEST[skip:illustrative response only]

Here you can see two significantly unique situations. Firstly, a node has a severely 
backed up write queue compared to other nodes. Secondly, a node shows historically 

> completed writes that are double any other node. These are both probably due 
> to either poorly distributed write-heavy indices or are a synergestic point 
> where multiple write-heavy indices overlap. Since primary and replica writes 

Reviewer: ```suggestion
completed writes that are double any other node. These are both probably due 
to either poorly distributed write-heavy indices, or to multiple write-heavy 
indices allocated to the same node. Since primary and replica writes 
```

are majorly the same amount of cluster work, we usually recommend setting 
<<total-shards-per-node,`index.routing.allocation.total_shards_per_node`>> 
to force index spreading after lining up index shard counts to total nodes. 

If it's non-obvious what indices are problematic, you can introspect further via 
<<indices-stats,the index stats API>> by running:

[source,console]
----
GET _stats?level=shards&human&expand_wildcards=all&filter_path=indices.*.total.indexing.index_total
----

> For more advanced analysis, going further you can poll for shard-level stats, 

Reviewer: ```suggestion
For more advanced analysis, you can poll for shard-level stats, 
```

though there's always link:{es-repo}/issues/17213[room for improvement]. 
A node may still experience hot spotting either due to write-heavy indices or by the 
overall shards it's hosting.

[discrete]
[[causes-shards-nodes]]
====== Node level

You can check for shard balancing via <<cat-allocation,cat allocation>>, though as of version 
8.6, <<modules-cluster,desired balancing>> may no longer fully expect to 
balance shards. Kindly note, both methods may temporarily show problematic imbalance during 
<<cluster-fault-detection,cluster stability issues>>.

For example, let's showcase two separate plausible issues using cat allocation:

[source,console]
----
GET _cat/allocation?v&s=node&h=node,shards,disk.percent,disk.indices,disk.used
----

Which could return:

[source,console-result]
----
node   shards disk.percent disk.indices disk.used
node_1    446           19      154.8gb   173.1gb
node_2     31           52       44.6gb   372.7gb
node_3    445           43      271.5gb   289.4gb
----
// TEST[skip:illustrative response only]

Here we see two significantly unique situations. `node_1` has 

> recently restarted, so it has a much lower number of shards than all other nodes. 

Author: @abdonpijpelink oops, this was suppose to reference `node_2`. I don't know how to edit my PR, can you help me fix it?

Support feedback would also like to add in this ballpark (about `node_2`'s situation) something to the effect of
```
While `node_2`'s shard count is low, it may become a index-write 
hot spot due with ongoing <<ilm-rollover,ILM Rollovers>>. This is a 
common root cause of write hot spots covered in the next section.
```

Reviewer: That's my mistake. I'll fix it and add that note.

----

[discrete]
[[causes-shards-index]]
====== Index level

Hot spotted nodes frequently surface via <<cat-thread-pool,cat thread pool>>'s 
`write` and `search` queue backups. For example:

[source,console]
----
GET _cat/thread_pool/write,search?v=true&s=n,nn&h=n,nn,q,a,r,c
----

Which could return:

[source,console-result]
----
n      nn       q a r    c
search node_1   3 1 0 1287
search node_2   0 2 0 1159
search node_3   0 1 0 1302
write  node_1 100 3 0 4259
write  node_2   0 4 0  980
write  node_3   1 5 0 8714
----
// TEST[skip:illustrative response only]

Here you can see two significantly unique situations. Firstly, `node_1` has a severely 
backed up write queue compared to other nodes. Secondly, `node_3` shows historically 
completed writes that are double any other node. These are both probably due 
to either poorly distributed write-heavy indices, or to multiple write-heavy 
indices allocated to the same node. Since primary and replica writes 
are majorly the same amount of cluster work, we usually recommend setting 
<<total-shards-per-node,`index.routing.allocation.total_shards_per_node`>> 
to force index spreading after lining up index shard counts to total nodes. 

> 

Author: @abdonpijpelink could you help me add two more common recoveries I missed (from support feedback); something to the effect:

```
We normally recommend heavy-write indices have sufficient primary 
`number_of_shards` and replica `number_of_replicas` to evenly spread 
across ingesting nodes. Alternatively, you can <<cluster-reroute,Reroute>> 
shards to more quiet nodes to alleviate the nodes with write hot spotting.
```
